subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c

```diff
@@ -0,0 +1,408 @@
+	}
+#else
+	char ver_string[strlen(VER_STRING_FMT2) +
+			strlen(BUILD_VERSION_STR) +
+			strlen(CDDL_VERSION)];
+
+	if (!err) {
+		snprintf(ver_string, sizeof(ver_string) - 1, VER_STRING_FMT2,
+			 BUILD_VERSION_STR, CDDL_VERSION);
+		ver_string[sizeof(ver_string) - 1] = '\0';
+	}
+#endif
+
+	LOG_INF("Request authorization with JWT");
+	err = nrf_cloud_coap_post("auth/jwt", err ? NULL : ver_string,
+				 (uint8_t *)jwt, strlen(jwt),
+				 COAP_CONTENT_FORMAT_TEXT_PLAIN, true, auth_cb, NULL);
+	if (err) {
+		return err;
+	}
+
+	if (authorized) {
+		LOG_INF("Authorized");
+
+		if (dtls_cid_is_active(sock)) {
+			err = dtls_session_save(sock);
+			if (!err) {
+				err = dtls_session_load(sock);
+				if (!err) {
+					LOG_INF("Saved DTLS CID session");
+				}
+			}
+		}
+	}
+	return err;
+}
+
+static K_SEM_DEFINE(cb_sem, 0, 1);
+
+struct user_cb {
+	coap_client_response_cb_t cb;
+	void *user_data;
+};
+
+static void client_callback(int16_t result_code, size_t offset, const uint8_t *payload, size_t len,
+			    bool last_block, void *user_data)
+{
+	struct user_cb *user_cb = (struct user_cb *)user_data;
+
+	LOG_DBG("Got callback: rc:%u.%02u, ofs:%u, lb:%u",
+		result_code / 32, result_code & 0x1f, offset, last_block);
+	if (payload && len) {
+		LOG_HEXDUMP_DBG(payload, MIN(len, 96), "payload received");
+	}
+	if (result_code == COAP_RESPONSE_CODE_UNAUTHORIZED) {
+		LOG_ERR("Device not authorized.  Reconnection required.");
+		authorized = false; /* Lost authorization; need to reconnect. */
+	}
+	if ((user_cb != NULL) && (user_cb->cb != NULL)) {
+		LOG_DBG("Calling user's callback %p", user_cb->cb);
+		user_cb->cb(result_code, offset, payload, len, last_block, user_cb->user_data);
+	}
+	if (last_block || (result_code >= COAP_RESPONSE_CODE_BAD_REQUEST)) {
+		LOG_DBG("Giving sem");
+		k_sem_give(&cb_sem);
+	}
+}
+
+static int client_transfer(enum coap_method method,
+			   const char *resource, const char *query,
+			   uint8_t *buf, size_t buf_len,
+			   enum coap_content_format fmt_out,
+			   enum coap_content_format fmt_in,
+			   bool response_expected,
+			   bool reliable,
+			   coap_client_response_cb_t cb, void *user)
+{
+	__ASSERT_NO_MSG(resource != NULL);
+	int err;
+	char path[MAX_COAP_PATH + 1];
+	struct user_cb user_cb = {
+		.cb = cb,
+		.user_data = user
+	};
+	struct coap_client_option options[1] = {{
+		.code = COAP_OPTION_ACCEPT,
+		.len = 1,
+		.value[0] = fmt_in
+	}};
+	struct coap_client_request request = {
+		.method = method,
+		.confirmable = reliable,
+		.path = path,
+		.fmt = fmt_out,
+		.payload = buf,
+		.len = buf_len,
+		.cb = client_callback,
+		.user_data = &user_cb
+	};
+
+	if (response_expected) {
+		request.options = options;
+		request.num_options = ARRAY_SIZE(options);
+	} else {
+		request.options = NULL;
+		request.num_options = 0;
+	}
+
+	strncpy(path, resource, MAX_COAP_PATH);
+	path[MAX_COAP_PATH] = '\0';
+	if (query) {
+		strncat(path, "?", MAX_COAP_PATH);
+		strncat(path, query, MAX_COAP_PATH);
```

Comment:
these 2 srtcats should use the remaining buffer size of path, not MAX_COAP_PATH.